code/controllers/subsystems/icon_updates.dm

```diff
@@ -0,0 +1,71 @@
+SUBSYSTEM_DEF(icon_update)
+	name = "Icon Updates"
+	wait = 1	// ticks
+	flags = SS_TICKER
+	priority = SS_PRIORITY_ICON_UPDATE
+	init_order = SS_INIT_ICON_UPDATE
+
+	// Linked lists, queue_refs[x] should have TRUE or args stored in queue_args[x]
+	var/list/queue_refs = list()	// Atoms
+	var/list/queue_args = list()	// TRUE or args
+
+/datum/controller/subsystem/icon_update/stat_entry()
+	..("Queue Refs: [queue_refs.len], Queue Args: [queue_args.len]")
+
+/datum/controller/subsystem/icon_update/Initialize()
+	fire(FALSE, TRUE)
+
+/datum/controller/subsystem/icon_update/fire(resumed = FALSE, no_mc_tick = FALSE)
+	if(!queue_refs.len)
+		suspend()
+		return
+
+	while (queue_refs.len)
+	
+		if(Master.map_loading)
+			return
+
+		// Pops the atom and it's args, while making sure
+		// to keep them in sync in case runtimes/badmin
+		var/atom/A = queue_refs[queue_refs.len]
+		var/myArgs = queue_args[queue_args.len]
+
+		queue_args.len = --queue_refs.len
+
+		if(QDELETED(A))
+			continue
+
+		if (islist(myArgs))
+			A.update_icon(arglist(myArgs))
+		else
+			A.update_icon()
+
+		if (no_mc_tick)
+			CHECK_TICK
+		else if (MC_TICK_CHECK)
+			return
+
+/atom
+	var/icon_update_queued = FALSE
+
+/atom/proc/queue_icon_update(...)
+	// Skips if this is already queued
+	if(!icon_update_queued)
+
+		icon_update_queued = TRUE
+		SSicon_update.queue_refs.Add(src)
+
+		// Makes sure these are in sync, in case runtimes/badmin
```

Comment:
If they're not the same length, the list is invalid anyways.

Reply:
Lowers the risk, and doesn't really change functionality, as queue_args len needs to be increased anyways. Looks clean too